subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskUpToDateIntegrationTest.groovy

```diff
@@ -50,22 +52,217 @@ class TaskUpToDateIntegrationTest extends AbstractIntegrationSpec {
+            class CustomTask extends DefaultTask {
+                @OutputFile
+                @Optional
+                File outputFile
+            
+                @TaskAction
+                void doAction() {
+                    if (outputFile != null) {
+                        outputFile.write "Task is running"
+                    }
+                }
+            }
+            
+            task customTask(type: CustomTask) {
+                outputFile = project.hasProperty('outputFile') ? file(project.property("outputFile")) : null
             }
         """
+
+        when:
+        succeeds "customTask"
+        then:
+        executedAndNotSkipped ":customTask"
+
+        when:
+        succeeds "customTask", "-PoutputFile=log.txt"
+        then:
+        executedAndNotSkipped ":customTask"
+    }
+
+    @Unroll
+    @Issue("https://github.com/gradle/gradle/issues/3073")
+    def "output files changed from #before to #after marks task #upToDateString"() {
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @OutputFiles
+                List<Object> outputFiles
+            
+                @TaskAction
+                void doAction() {
+                    outputFiles.each { output ->
+                         def outputFile = output.call()
+                         if (outputFile != null) {
+                            outputFile.text = "task executed"
+                         }
+                    }
+                }
+            }
+            
+            def lazyProperty(String name) {
+                def value = project.findProperty(name)
+                def outputFile = value ? file(value) : null
+                return { -> outputFile }
+            }
+            
+            task customTask(type: CustomTask) {
+                outputFiles = [lazyProperty('output0'), lazyProperty('output1'), lazyProperty('output2')]
+            }
+        """
+
+        when:
+        succeeds customTaskWithOutputs(before)
+        then:
+        executedAndNotSkipped ":customTask"
+
+        when:
+        succeeds customTaskWithOutputs(after)
+        then:
+        if (upToDate) {
+            skipped ":customTask"
+        } else {
+            executedAndNotSkipped ":customTask"
+        }
+
+        where:
+        before                                        | after                          | upToDate
+        [null, 'outputFile']                          | ['outputFile', null]           | true
+        ['outputFile1', null]                         | ['outputFile1', 'outputFile2'] | false
+        ['outputFile1', 'outputFile2', 'outputFile3'] | ['outputFile1', 'outputFile2'] | false
+        ['outputFile1', 'outputFile2']                | ['outputFile2', 'outputFile3'] | false
+        ['outputFile1', 'outputFile2']                | []                             | false
+        ['outputFile1', 'outputFile2']                | ['outputFile1']                | false
```

Comment:
There should be another test case where we add one more input. Maybe even two cases:

* `['outputFile1']` -> `['outputFile1', 'outputFile2']`
* `[null, 'outputFile1']` -> `['outputFile1', 'outputFile2']`

Reply:
I added the cases.